### PR TITLE
schemahcl: reduce eval mem usage

### PR DESCRIPTION
**Pull request by a8m**
```
benchmark              old ns/op     new ns/op     delta
BenchmarkSchema-16     569319394     238469401     -58.11%

benchmark              old allocs     new allocs     delta
BenchmarkSchema-16     1327927        1174085        -11.59%

benchmark              old bytes     new bytes     delta
BenchmarkSchema-16     293956792     186829161     -36.44%
```

### CODE REVIEW

File: schemahcl/context.go [hunk 1]
```diff
@@ -402,13 +409,11 @@ func blocksOfType(blocks hclsyntax.Blocks, typeName string) []*hclsyntax.Block {
...
 func attrMap(attrs hclsyntax.Attributes) map[string]cty.Value {
-	out := make(map[string]cty.Value)
+	out := make(map[string]cty.Value, len(attrs))
 	for _, v := range attrs {
-		value, diag := v.Expr.Value(nil)
-		if diag.HasErrors() {
-			continue
+		if value, diag := v.Expr.Value(nil); !diag.HasErrors() {
+			out[v.Name] = value
 		}
```
Comment (masseelch) [inline schemahcl/context.go:416]: What kind of perf impact has this?

File: schemahcl/extension.go [hunk 2]
```diff
@@ -547,10 +547,16 @@ func scanAttr(key string, r *Resource, field reflect.Value) error {
 	return nil
 }
 
+// specCache holds a cache for type-spec fields descriptions.
+var specCache sync.Map
+
```
Comment (masseelch) [inline schemahcl/extension.go:552]: Is this called concurrently? A simple map is faster.

File: schemahcl/schemahcl.go [hunk 3]
```diff
@@ -479,28 +480,11 @@ func (s *State) mayScopeContext(ctx *hcl.EvalContext, scope []string) *hcl.EvalC
...
-	// Override the parent context with the scoped variables and functions.
-	for n, v := range vars {
-		nctx.Variables[n] = v
-	}
-	for n, f := range funcs {
-		nctx.Functions[n] = f
-	}
-	// A patch from the past. Should be moved
-	// to specific scopes in the future.
-	nctx.Functions["sql"] = rawExprFunc
```
Comment (masseelch) [inline schemahcl/schemahcl.go:502]: Where is this? No longer needed since we don't copy to a new map?

File: sql/postgres/sqlspec.go [hunk 4]
```diff
@@ -900,7 +900,6 @@ var TypeRegistry = schemahcl.NewRegistry(
 		schemahcl.NewTypeSpec(TypeDateRange),
 		schemahcl.NewTypeSpec(TypeDateMultiRange),
 		schemahcl.NewTypeSpec("hstore"),
-		schemahcl.NewTypeSpec("sql", schemahcl.WithAttributes(&schemahcl.TypeAttr{Name: "def", Required: true, Kind: reflect.String})),
```
Comment (masseelch) [inline sql/postgres/sqlspec.go:903]: Why remove?

File: schemahcl/context.go (hunk 1 above)
Comment (a8m) [reply schemahcl/context.go:416]: no perf. just avoiding the unnecessary ifs.

File: schemahcl/schemahcl.go (hunk 3 above)
Comment (a8m) [reply schemahcl/schemahcl.go:502]: global definitions (see stdFuncs).

File: schemahcl/extension.go (hunk 2 above)
Comment (a8m) [reply schemahcl/extension.go:552]: Yes - our tests :D

**Pull request merged**